janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexOptimizerStrategy.java

```diff
@@ -90,37 +82,36 @@ private OptimizablePosition getOptimizablePosition(T originalStep) {
         return OptimizablePosition.NONE;
     }
 
-    private void replaceSequence(T originalStep, OptimizablePosition pos,
-                                   P<?> predicate) {
+    private void replaceSequence(T originalStep, OptimizablePosition pos) {
         switch (pos) {
         case V2E_E2V_ID:
-            replaceSequenceV2EthenE2VthenID(originalStep, predicate);
+            replaceSequenceV2EthenE2VthenID(originalStep);
             break;
         case V2V_ID:
-            replaceSequenceV2VthenID(originalStep, predicate);
+            replaceSequenceV2VthenID(originalStep);
             break;
         default:
             break;
         }
     }
 
-    private void replaceSequenceV2EthenE2VthenID(T originalStep, P<?> predicate) {
+    private void replaceSequenceV2EthenE2VthenID(T originalStep) {
         Traversal.Admin<?,?> traversal = originalStep.getTraversal();
 
         Step<Edge,Vertex> e2vStep;
         e2vStep = (Step<Edge, Vertex>) originalStep.getPreviousStep();
 
         // create new has("~adjacent", id_value) step before e2v step
-        HasStep<Edge> hasAdjacentIdStep = makeHasAdjacentIdStep(traversal, predicate);
+        FilterStep<Edge> filterByAdjacentIdStep = makeFilterByAdjacentIdStep(traversal, originalStep);
 
         // remove original selection step
         traversal.removeStep(originalStep);
```

Comment:
First of all, I don't know where I would get the exact step position from, in order to use `removeStep(final int index)`. My only idea would be to iterate all steps myself. 
And in addition, I'm not sure why it would be better to delete the old step beforehand. Notice that the new step is not added to the query at this point so it makes no difference in comparing all steps of the traversal.

Reply:
I agree at the moment, putting this step beforehand makes no difference. What I meant was:

We know that `removeStep(originalStep)` essentially traverses all steps and compares the steps with originalStep using `equals()` method. We have to ensure originalStep is not modified after we extract it from the traversal, otherwise `equals()` will always return false.

In the last operation, `FilterStep<Edge> filterByAdjacentIdStep = makeFilterByAdjacentIdStep(traversal, originalStep);`, the `originalStep` is passed in as a parameter. Within that function, `originalStep` is added to a new ad-hoc traversal. This means, `originalStep` is in fact shared by the original traversal and the nested ad-hoc traversal. As discussed above, we have to make sure this `originalStep` is not mutated in the nested ad-hoc traversal, otherwise it cannot be correctly removed from the original traversal afterward. This sounds a bit risky to me because I don't see a way to enforce this "contract".

Using an analogy below, this is a competition between
```java
List<Object> list = new ArrayList<>(Arrays.asList(obj1, obj2));
runFunc(obj1);    // a function call that arguably might mutate obj1
list.remove(obj1);
```

vs


```java
List<Object> list = new ArrayList<>(Arrays.asList(obj1, obj2));
list.remove(obj1);
runFunc(obj1);
```

Reply:
Now I understand you and what you say makes perfect sense. Thanks for clarification!